### PR TITLE
workload/schemachange: enhance logging to be more consumable

### LINKED ISSUE #89284: schemachange/random-load: improve error message output
We recently had a failure here: https://github.com/cockroachdb/cockroach/issues/89146
The blurb in the issue description is not helpful. One has to dig into `https://teamcity.cockroachdb.com/repository/download/Cockroach_Nightlies_RoachtestNightlyGceBazel/6731773:id/schemachange/random-load/run_1/artifacts.zip!/run_141003.097639402_n1_workload_run_schemachange.log`  to find `UNEXPECTED ERROR; Failed to generate a random operation` in a jumble of mixed formatting which is very hard to make sense of. 

Jira issue: CRDB-20190

### PR DESCRIPTION
**Pull request by fqazi**
These changes modify the logging inside the schema changer workload to do the following:

Fixes: #89284

1) Generate more easily consumable error state information that is kept was JSON. This works by adding an new error wrapper to help propagate errors.
2) Turn the opGenLog into a structured log for more easy consume ability.
3) Eliminate unnecessary fields in the output by omitting empty optional fields from the JSON log.
4) Eliminate unnecessary state information dumps, where successful operations were also dumping this information out.

Release note: None

As an example, the new logging during a failure in the JSON format we currently generate the following output on master:
```
{
  "workerId": 0,
  "clientTimestamp": "03:23:30.14084",
  "ops": [
    "BEGIN",
    "QUERY: SELECT 'validating all objects', crdb_internal.validate_multi_region_zone_configs(), Expected Errors: , Potential Errors: ",
    "QUERY: CREATE TABLE public.table8 (col8_9 GEOMETRY NOT NULL, col8_10 BOX2D NOT NULL, col8_11 FLOAT4 NOT NULL, col8_12 TIMETZ NOT NULL, col8_13 \"char\" NOT NULL, col8_14 NAME NOT NULL, col8_15 DECIMAL NOT NULL, col8_16 INET NOT NULL, col8_17 INT8 NO
T NULL, col8_18 STRING NOT NULL AS (lower(col8_13)) STORED, col8_19 FLOAT8 NOT NULL AS (col8_11 + 0.05183445289731026:::FLOAT8) VIRTUAL, PRIMARY KEY (col8_10, col8_18 DESC, col8_17 DESC, col8_19, col8_15 DESC, col8_11 ASC, col8_13 DESC, col8_14, col8_16
DESC), UNIQUE (col8_17 ASC, col8_12 DESC, col8_19, col8_18) PARTITION BY LIST (col8_17, col8_12, col8_19, col8_18) (PARTITION table8_part_0 VALUES IN ((1112324545982352417:::INT8, '08:25:58.380905+00:58:00':::TIMETZ, 2.308196400279898:::FLOAT8, e' >\\t\\
bp\\x02\\x047':::STRING)), PARTITION table8_part_1 VALUES IN (((-4169758796321653179):::INT8, '12:59:18.130563-10:24:00':::TIMETZ, (-0.7093454254464202):::FLOAT8, e'k\\x1a#`!#c\\x7fY':::STRING)), PARTITION table8_part_2 VALUES IN (((-5076717300105623401)
:::INT8, '07:12:07.412407-07:06:00':::TIMETZ, (-0.7914235188323253):::FLOAT8, e'KIn3#\\rAk':::STRING)), PARTITION table8_part_3 VALUES IN ((2147483647:::INT8, '18:07:00.223705+03:00:00':::TIMETZ, 1.4698528925964642:::FLOAT8, e'S\\x047\\x03\\x02VW\\x0e'::
:STRING)), PARTITION table8_part_4 VALUES IN (((-5362989756654691582):::INT8, '01:31:00.271577+14:32:00':::TIMETZ, (-0.3369973678230378):::FLOAT8, '/fKW':::STRING)), PARTITION table8_part_5 VALUES IN (((-6629264169873001169):::INT8, '02:01:58.855859-02:5
8:00':::TIMETZ, (-0.8050523865991726):::FLOAT8, '':::STRING)), PARTITION table8_part_6 VALUES IN (((-5722173834231977500):::INT8, '17:00:12.308808-06:17:00':::TIMETZ, (-0.9479345049607424):::FLOAT8, '':::STRING)), PARTITION table8_part_7 VALUES IN (((-83
7799042489541168):::INT8, '06:37:23.35124+07:43:00':::TIMETZ, (-1.0469263893545366):::FLOAT8, e'\\x03\\x01l(n\\x1cv9':::STRING)), PARTITION table8_part_8 VALUES IN ((3264035864940073454:::INT8, '11:23:56.130029+12:36:00':::TIMETZ, 0.6868497639211962:::FL
OAT8, '':::STRING)), PARTITION \"DEFAULT\" VALUES IN ((DEFAULT, DEFAULT, DEFAULT, DEFAULT))), INDEX (lower(CAST(col8_9 AS STRING)) ASC, lower(CAST(col8_16 AS STRING))), INDEX (col8_15 ASC, col8_13 ASC, col8_19, col8_16 ASC, lower(CAST(col8_16 AS STRING))
 ASC, col8_10, col8_14 ASC, col8_18 ASC), INDEX (lower(col8_14) DESC, col8_13 DESC, col8_14 ASC), INDEX (col8_19, (col8_11 + (-0.5048524141311646):::FLOAT8) ASC, col8_18, col8_13, lower(CAST(col8_10 AS STRING)) DESC, col8_12 ASC) STORING (col8_9) NOT VIS
IBLE, UNIQUE (col8_18 ASC, col8_19, col8_15 ASC, col8_10, col8_14, col8_17 ASC, col8_16 ASC) STORING (col8_9, col8_12), FAMILY (col8_18, col8_16, col8_13), FAMILY (col8_11, col8_12, col8_15), FAMILY (col8_10), FAMILY (col8_14), FAMILY (col8_9), FAMILY (c
ol8_17)), Expected Errors: , Potential Errors: "
  ],
  "expectedExecErrors": "",
  "expectedCommitErrors": "",
  "message": "***UNEXPECTED ERROR; Failed to generate a random operation\n OpGen log: \nQUERY [\"SELECT EXISTS (\\n\\tSELECT table_name\\n    FROM information_schema.tables \\n   WHERE table_schema = $1\\n     AND table_name = $2\\n   )\" [\"public\" \"t
able8\"]] :true\nStmts: \n[QUERY: SELECT 'validating all objects', crdb_internal.validate_multi_region_zone_configs(), Expected Errors: , Potential Errors: QUERY: CREATE TABLE public.table8 (col8_9 GEOMETRY NOT NULL, col8_10 BOX2D NOT NULL, col8_11 FLOAT
4 NOT NULL, col8_12 TIMETZ NOT NULL, col8_13 \"char\" NOT NULL, col8_14 NAME NOT NULL, col8_15 DECIMAL NOT NULL, col8_16 INET NOT NULL, col8_17 INT8 NOT NULL, col8_18 STRING NOT NULL AS (lower(col8_13)) STORED, col8_19 FLOAT8 NOT NULL AS (col8_11 + 0.051
83445289731026:::FLOAT8) VIRTUAL, PRIMARY KEY (col8_10, col8_18 DESC, col8_17 DESC, col8_19, col8_15 DESC, col8_11 ASC, col8_13 DESC, col8_14, col8_16 DESC), UNIQUE (col8_17 ASC, col8_12 DESC, col8_19, col8_18) PARTITION BY LIST (col8_17, col8_12, col8_1
9, col8_18) (PARTITION table8_part_0 VALUES IN ((1112324545982352417:::INT8, '08:25:58.380905+00:58:00':::TIMETZ, 2.308196400279898:::FLOAT8, e' >\\t\\bp\\x02\\x047':::STRING)), PARTITION table8_part_1 VALUES IN (((-4169758796321653179):::INT8, '12:59:18
.130563-10:24:00':::TIMETZ, (-0.7093454254464202):::FLOAT8, e'k\\x1a#`!#c\\x7fY':::STRING)), PARTITION table8_part_2 VALUES IN (((-5076717300105623401):::INT8, '07:12:07.412407-07:06:00':::TIMETZ, (-0.7914235188323253):::FLOAT8, e'KIn3#\\rAk':::STRING)),
 PARTITION table8_part_3 VALUES IN ((2147483647:::INT8, '18:07:00.223705+03:00:00':::TIMETZ, 1.4698528925964642:::FLOAT8, e'S\\x047\\x03\\x02VW\\x0e':::STRING)), PARTITION table8_part_4 VALUES IN (((-5362989756654691582):::INT8, '01:31:00.271577+14:32:00
':::TIMETZ, (-0.3369973678230378):::FLOAT8, '/fKW':::STRING)), PARTITION table8_part_5 VALUES IN (((-6629264169873001169):::INT8, '02:01:58.855859-02:58:00':::TIMETZ, (-0.8050523865991726):::FLOAT8, '':::STRING)), PARTITION table8_part_6 VALUES IN (((-57
22173834231977500):::INT8, '17:00:12.308808-06:17:00':::TIMETZ, (-0.9479345049607424):::FLOAT8, '':::STRING)), PARTITION table8_part_7 VALUES IN (((-837799042489541168):::INT8, '06:37:23.35124+07:43:00':::TIMETZ, (-1.0469263893545366):::FLOAT8, e'\\x03\\
x01l(n\\x1cv9':::STRING)), PARTITION table8_part_8 VALUES IN ((3264035864940073454:::INT8, '11:23:56.130029+12:36:00':::TIMETZ, 0.6868497639211962:::FLOAT8, '':::STRING)), PARTITION \"DEFAULT\" VALUES IN ((DEFAULT, DEFAULT, DEFAULT, DEFAULT))), INDEX (lo
wer(CAST(col8_9 AS STRING)) ASC, lower(CAST(col8_16 AS STRING))), INDEX (col8_15 ASC, col8_13 ASC, col8_19, col8_16 ASC, lower(CAST(col8_16 AS STRING)) ASC, col8_10, col8_14 ASC, col8_18 ASC), INDEX (lower(col8_14) DESC, col8_13 DESC, col8_14 ASC), INDEX
 (col8_19, (col8_11 + (-0.5048524141311646):::FLOAT8) ASC, col8_18, col8_13, lower(CAST(col8_10 AS STRING)) DESC, col8_12 ASC) STORING (col8_9) NOT VISIBLE, UNIQUE (col8_18 ASC, col8_19, col8_15 ASC, col8_10, col8_14, col8_17 ASC, col8_16 ASC) STORING (c
ol8_9, col8_12), FAMILY (col8_18, col8_16, col8_13), FAMILY (col8_11, col8_12, col8_15), FAMILY (col8_10), FAMILY (col8_14), FAMILY (col8_9), FAMILY (col8_17)), Expected Errors: , Potential Errors:]\n: Blah"
}
{
```

With these changes, we will get more readable output (note the scenarios are not same, but we can see readability in general from them):
```
{
  "workerId": 0,
  "clientTimestamp": "03:19:26.209338",
  "ops": [
    "BEGIN",
    {
      "sql": "SELECT t1.col47_54 AS col0,t1.col47_52 AS col1,t1.col47_50 AS col2,t1.col47_56 AS col3,t1.col47_53 AS col4,t1.col47_55 AS col5 FROM public.table57 AS t0 ,public.table47 AS t1  FETCH FIRST 1 ROWS ONLY"
    }
  ],
  "expectedExecErrors": "",
  "expectedCommitErrors": "",
  "message": "***UNEXPECTED ERROR; Received an unexpected execution error.: ERROR: relation \"public.table57\" does not exist (SQLSTATE 42P01)",
  "errorState": {
    "queriesForGeneratingErrors": [
      {
        "query": "SELECT EXISTS (SELECT schema_name FROM information_schema.schemata WHERE schema_name = $1)",
        "queryArgs": [
          "public"
        ],
        "result": true
      },
      {
        "query": "SELECT EXISTS (SELECT sequence_name FROM information_schema.sequences WHERE (sequence_schema = $1) AND (sequence_name = $2))",
        "queryArgs": [
          "public",
          "seq1"
        ],
        "result": false
      },
      {
        "query": "SELECT EXISTS (SELECT schema_name FROM information_schema.schemata WHERE schema_name = $1)",
        "queryArgs": [
          "public"
        ],
        "result": true
      },
      {
        "query": "SELECT EXISTS (SELECT * FROM [SHOW REGIONS FROM DATABASE])",
        "result": false
      },
      {
        "query": "SELECT EXISTS (SELECT table_name FROM information_schema.tables WHERE (table_schema = $1) AND (table_name = $2))",
        "queryArgs": [
          "public",
          "table5"
        ],
        "result": false
      },
      {
        "query": "SELECT EXISTS (SELECT schema_name FROM information_schema.schemata WHERE schema_name = $1)",
        "queryArgs": [
          "public"
        ],
        "result": true
      },
      {
        "query": "SELECT EXISTS (SELECT table_name FROM information_schema.tables WHERE (table_schema = $1) AND (table_name = $2))",
        "queryArgs": [
          "public",
          "table5"
        ],
        "result": true
      },
      {
        "query": "SELECT EXISTS (SELECT table_name FROM information_schema.tables WHERE (table_schema = $1) AND (table_name = $2))",
        "queryArgs": [
          "public",
          "table5"
        ],
        "result": true
      },
      {
        "query": "SELECT EXISTS (SELECT table_name FROM information_schema.tables WHERE (table_schema = $1) AND (table_name = $2))",
        "queryArgs": [
          "public",
          "table5"
        ],
        "result": true
      },
      {
        "query": "SELECT EXISTS (SELECT schema_name FROM information_schema.schemata WHERE schema_name = $1)",
        "queryArgs": [
          "public"
        ],
        "result": true
      },
      {
        "query": "SELECT EXISTS (SELECT table_name FROM information_schema.tables WHERE (table_schema = $1) AND (table_name = $2))",
        "queryArgs": [
          "public",
          "table13"
        ],
        "result": false
      },
      {
        "query": "SELECT EXISTS (SELECT sequence_name FROM information_schema.sequences WHERE (sequence_schema = $1) AND (sequence_name = $2))",
        "queryArgs": [
          "public",
          "seq1"
        ],
        "result": true
      },
      {
        "query": "SELECT EXISTS (SELECT table_name FROM information_schema.tables WHERE (table_schema = $1) AND (table_name = $2))",
        "queryArgs": [
          "public",
          "table14"
        ],
        "result": false
      },
      {
        "query": "SELECT EXISTS (SELECT schema_name FROM information_schema.schemata WHERE schema_name = $1)",
        "queryArgs": [
          "public"
        ],
        "result": true
      },
      {
        "query": "SELECT EXISTS (SELECT sequence_name FROM information_schema.sequences WHERE (sequence_schema = $1) AND (sequence_name = $2))",
        "queryArgs": [
          "public",
          "seq15"
        ],
        "result": false
      },
      {
        "query": "SELECT EXISTS (SELECT sequence_name FROM information_schema.sequences WHERE (sequence_schema = $1) AND (sequence_name = $2))",
        "queryArgs": [
          "public",
          "seq15"
        ],
        "result": true
      },
      {
        "query": "SELECT EXISTS (SELECT table_name FROM information_schema.tables WHERE (table_schema = $1) AND (table_name = $2))",
        "queryArgs": [
          "public",
          "table39"
        ],
        "result": false
      },
      {
        "query": "SELECT EXISTS (SELECT sequence_name FROM information_schema.sequences WHERE (sequence_schema = $1) AND (sequence_name = $2))",
        "queryArgs": [
          "public",
          "seq42"
        ],
        "result": false
      },
      {
        "query": "SELECT EXISTS (SELECT table_name FROM information_schema.tables WHERE (table_schema = $1) AND (table_name = $2))",
        "queryArgs": [
          "public",
          "table43"
        ],
        "result": false
      },
      {
        "query": "SELECT EXISTS (SELECT schema_name FROM information_schema.schemata WHERE schema_name = $1)",
        "queryArgs": [
          "schema41"
        ],
        "result": true
      },
      {
        "query": "WITH database_id AS (SELECT id FROM system.namespace WHERE ((\"parentID\" = 0) AND (\"parentSchemaID\" = 0)) AND (name = current_database())), schema_id AS (SELECT nsp.id FROM system.namespace AS nsp JOIN database_id ON ((\"parentID\" = database_id.id) AND (\"parentSchemaID\" = 0)) AND (name = $1)), descriptor_ids AS (SELECT nsp.id FROM system.namespace AS nsp, schema_id, database_id WHERE (nsp.\"parentID\" = database_id.id) AND (nsp.\"parentSchemaID\" = schema_id.id)), descriptors AS (SELECT crdb_internal.pb_to_json('cockroach.sql.sqlbase.Descriptor', descriptor) AS descriptor FROM system.descriptor AS descriptors JOIN descriptor_ids ON descriptors.id = descriptor_ids.id), types AS (SELECT descriptor FROM descriptors WHERE (descriptor->'type') IS NOT NULL), table_references AS (SELECT json_array_elements((descriptor->'table')->'dependedOnBy') AS ref FROM descriptors WHERE (descriptor->'table') IS NOT NULL), dependent AS (SELECT (ref->>'id')::INT8 AS id FROM table_references), referenced_descriptors AS (SELECT json_array_elements_text((descriptor->'type')->'referencingDescriptorIds')::INT8 AS id FROM types) SELECT EXISTS (SELECT * FROM system.namespace WHERE ((id IN (SELECT id FROM referenced_descriptors)) AND (\"parentSchemaID\" NOT IN (SELECT id FROM schema_id))) AND (id NOT IN (SELECT id FROM dependent)))",
        "queryArgs": [
          "schema41"
        ],
        "result": false
      },
      {
        "query": "SELECT EXISTS (SELECT schema_name FROM information_schema.schemata WHERE schema_name = $1)",
        "queryArgs": [
          "public"
        ],
        "result": true
      },
      {
        "query": "SELECT EXISTS (SELECT * FROM [SHOW REGIONS FROM DATABASE])",
        "queryArgs": null,
        "result": false
      },
      {
        "query": "SELECT EXISTS (SELECT table_name FROM information_schema.tables WHERE (table_schema = $1) AND (table_name = $2))",
        "queryArgs": [
          "public",
          "table47"
        ],
        "result": false
      },
      {
        "query": "SELECT EXISTS (SELECT schema_name FROM information_schema.schemata WHERE schema_name = $1)",
        "queryArgs": [
          "public"
        ],
        "result": true
      },
      {
        "query": "SELECT EXISTS (SELECT table_name FROM information_schema.tables WHERE (table_schema = $1) AND (table_name = $2))",
        "queryArgs": [
          "public",
          "table57"
        ],
        "result": false
      },
      {
        "query": "SELECT EXISTS (SELECT table_name FROM information_schema.tables WHERE (table_schema = $1) AND (table_name = $2))",
        "queryArgs": [
          "public",
          "table47"
        ],
        "result": true
      }
    ],
    "previousStatements": [
      "SELECT t1.col47_54 AS col0,t1.col47_52 AS col1,t1.col47_50 AS col2,t1.col47_56 AS col3,t1.col47_53 AS col4,t1.col47_55 AS col5 FROM public.table57 AS t0 ,public.table47 AS t1  FETCH FIRST 1 ROWS ONLY"
    ]
  }
}
````

### CODE REVIEW

**Comment by fqazi** [comment]
@postamar Added an example snippet for an error scenario, there is also some cruft reduction in successful cases to make things readable.

**Comment by postamar** [comment]
Thanks for the examples, Faizan. This is obviously a significant improvement. 

Would it be possible to go a bit further and clean up the `query` fields in `queriesForGeneratingErrors` by stripping the escaped newlines, etc? Also split up that field into `query` and `args` to expose its underlying structure as a prepared statement. I'm really not a huge fan of how `LogQueryResults` takes a `fmt.Sprintf("%q %q", query, args)` and this seems like a simple-enough change.

Basically it would be really nice if we could copy-paste the strings in the json log into sqlfmpt with minimal post-processing.

**Comment by fqazi** [comment]
@postamar Yes let me modify this to be pretty printed and a similar thing with the results so that types are properly outputted in the JSON, instead of always being strings

**Comment by fqazi** [comment]
@postamar This should hopefully be more consumable, updated the PR description to include the new output

File: pkg/workload/schemachange/operation_generator.go [hunk 1]
```diff
@@ -2593,23 +2627,41 @@ func makeOpStmt(queryType opStmtType) *opStmt {
...
-		og.potentialCommitErrors.String(),
-		og.GetOpGenLog(),
-		og.stmtsInTxt)
+// ErrorState wraps schemachange workload errors to have state information for
+// the purpose of dumping in our JSON log.
+type ErrorState struct {
+	cause                      error
+	ExpectedErrors             []string      `json:"expectedErrors,omitempty"`
+	PotentialErrors            []string      `json:"potentialErrors,omitempty"`
+	ExpectedCommitErrors       []string      ` json:"expectedCommitErrors,omitempty"`
```
Comment (postamar) [inline pkg/workload/schemachange/operation_generator.go:2636]: nit: whitespace following `

File: pkg/workload/schemachange/schemachange.go [hunk 2]
```diff
@@ -573,6 +582,30 @@ func (l *logger) addExpectedErrors(execErrors errorCodeSet, commitErrors errorCo
...
+		if l.currentLogEntry.mu.entry != nil {
+			var state *ErrorState
+			if errors.As(err, &state) {
+				l.currentLogEntry.mu.entry.ErrorState = state
+			}
+		}
+	}()
+
+	l.flushLogAndLock(tx, err.Error(), true)
+	l.currentLogEntry.mu.Unlock()
```
Comment (postamar) [inline pkg/workload/schemachange/schemachange.go:606]: Is this intentional? Looks like you're unlocking twice.

**Review by postamar — APPROVED** [review]
Thanks, I like this!

File: pkg/workload/schemachange/schemachange.go (hunk 2 above)
Comment (fqazi) [reply pkg/workload/schemachange/schemachange.go:606]: Yes, the flushLogAndLock holds onto the lock. Not a fan of this design, I'll poke at it with a later PR.

**Comment by fqazi** [comment]
@postamar TFTR!

bors r+
